src/components/AqiGraph/AqiGraph.tsx

```diff
@@ -0,0 +1,311 @@
+              if (element.timestamp) {
+                const date = element.timestamp.toDate();
+                const hour =
+                  date.getHours() + date.getMinutes() / minutesPerHour;
+                let hoursAgo = currentHour - hour;
+                // If the reading is from yesterday
+                if (date.getDay() !== currentDay) {
+                  hoursAgo = currentHour + (hoursPerDay - hour);
+                }
+                // Limit range to [0,24]
+                hoursAgo = Math.max(Math.min(hoursAgo, hoursPerDay), 0);
+                if (element.aqi > maxAqi) {
+                  maxAqi = element.aqi;
+                }
+                const newElement: GraphElement = {x: hoursAgo, y: element.aqi};
+                // Add element to data
+                if (element.aqi <= good) {
+                  allData.good.push(newElement);
+                } else if (element.aqi < moderate) {
+                  allData.moderate.push(newElement);
+                } else if (element.aqi < sensitiveGroups) {
+                  allData.sensitive.push(newElement);
+                } else if (element.aqi < unhealthy) {
+                  allData.unhealthy.push(newElement);
+                } else if (element.aqi < veryUnhealthy) {
+                  allData.veryUnhealthy.push(newElement);
+                } else {
+                  allData.hazardous.push(newElement);
+                }
+              }
+            }
+            setYAxisLimit(Math.max(maxAqi, defaultYLimit));
+            setData(allData);
+          }
+        }
+      });
+    }
+  }, [sensorDocId]);
+
+  // The graph doesn't include the hazardous category unless thee are data points
+  // that fall in that category. Any time the maximum y-axis value changes, this effect
+  // updates the styling for the graph accordingly
+  useEffect(() => {
+    if (yAxisLimit === defaultYLimit) {
+      // Don't include hazardous color and extra grid line
+      /* eslint-disable-next-line no-magic-numbers */
+      setYAxisTicks([0, 50, 100, 150, 200, 300]);
+      setHorizontalFill([
+        '#8F3F97',
+        '#8F3F97',
+        '#FF0202',
+        '#FF7E02',
+        '#FEFF00',
+        '#08E400',
+      ]);
+    } else {
+      // Include hazardous color and extra grid line
+      /* eslint-disable-next-line no-magic-numbers */
+      setYAxisTicks([0, 50, 100, 150, 200, 300, yAxisLimit]);
+      setHorizontalFill([
+        '#7E0224',
+        '#7E0224',
+        '#8F3F97',
+        '#FF0202',
+        '#FF7E02',
+        '#FEFF00',
+        '#08E400',
+      ]);
+    }
+  }, [yAxisLimit]);
+
+  const formatLabels = (hoursAgo: number): string => {
+    const weekdays = [
+      t('Monday'),
+      t('Tuesday'),
+      t('Wednesday'),
+      t('Thursday'),
+      t('Friday'),
+      t('Saturday'),
+      t('Sunday'),
+    ];
+    const hoursPerPeriod = 12;
+    const sunday = 6;
+    // Get the local time from the user's browser
+    const date = new Date();
```

Comment:
I forget, will the Firestore queries automatically handle time zone conversion for us?

Reply:
I use `getDay()`, `getHours()`, and `getMinutes()` which get the those values in the **time zone of the browser**. So all the axis labels are in the local timezone. I also use these functions when I calculate how many hours ago each data point is from, so throughout this code I'm using the user's local time (even though it doesn't really matter for the hours-ago calculation above).